### PR TITLE
feat(test-runner-visual-regression): add testName to GetNameArgs

### PR DESCRIPTION
**Pull request by WickyNilliams**
by supplying test name to each of the extensibility points, it becomes easier to store images relative to the test itself.

For example, this will store snapshot images in a directory adjacent to the test file:

```js
visualRegressionPlugin({
  getBaselineName: ({ name, browser, testFile }) =>
    path.join(testFile, "..", "screenshots", browser, "baseline", name),

  getDiffName: ({ browser, name, testFile }) =>
    path.join(testFile, "..", "screenshots", browser, "failed", `${name}-diff`),
    
  getFailedName: ({ browser, name, testFile }) =>
    path.join(testFile, "..", "screenshots", browser, "failed", name),
});

```

## What I did

1. Added `testFile` property to `GetNameArgs`
2. Passed `session.testFile` from plugin through to extensibility points like `getBaselineName` etc
3. Updated README

### CODE REVIEW

File: packages/test-runner-visual-regression/src/visualRegressionPlugin.ts
Comment (WickyNilliams) [inline packages/test-runner-visual-regression/src/visualRegressionPlugin.ts:75]: Feels like the argument list is getting a little unwieldy here, but I just made the smallest addition I could. Didn't want to start restructuring things!

Comment (web-padawan) [reply packages/test-runner-visual-regression/src/visualRegressionPlugin.ts:75]: Thanks for the reasonable approach. I think you have a valid point here about the argument list.

What if we change `visualDiffCommand` to accept `context` object as a third argument?
It could contain `browser` and `testFile` fields for now and maybe something else later. 

In that case we migth want to also export `VisuallDiffCommandContext` interface.

**Review by web-padawan — COMMENTED** [review]
In general the approach looks good to me 👍   By the way, I was previously asking for this in #1040

Please consider implementing my suggestion about changing visualDiff command arguments.
And also please add a changeset for this PR using `npx changeset` to trigger a version bump.

If you think that my suggestion does make sense, I think the bump cab be minor. But technically it seems to be a non-breaking change so it can be a patch during 0.x

File: packages/test-runner-visual-regression/src/visualRegressionPlugin.ts
Comment (web-padawan) [reply packages/test-runner-visual-regression/src/visualRegressionPlugin.ts:75]: Sorry, I have minimized the above comment because it appears to be shown two times for me 😅

**Review by web-padawan — CHANGES_REQUESTED** [review]
Please rebase as there are merge conflicts now, also please check my comment.

**Comment by WickyNilliams** [comment]
@web-padawan yeah i'll rebase now and make the changes you suggested

File: packages/test-runner-visual-regression/src/visualRegressionPlugin.ts
Comment (WickyNilliams) [reply packages/test-runner-visual-regression/src/visualRegressionPlugin.ts:75]: haha no worries. I am thinking of also adding the `payload.name` into the new Context type, since it is used everywhere browser and testFile are used? e.g. as it is now:

```js

options.getBaselineName({ browser, name, testFile })
```

could become:

```js
options.getBaselineName(context)
```

what do you think?

Comment (WickyNilliams) [reply packages/test-runner-visual-regression/src/visualRegressionPlugin.ts:75]: Though by then the type is identical to `GetNameArgs` haha. Seems like there's some overlap in intent

Comment (WickyNilliams) [reply packages/test-runner-visual-regression/src/visualRegressionPlugin.ts:75]: Maybe it's best to stick with your initial suggestion

**Comment by WickyNilliams** [comment]
Added changeset, made changes you suggested, rebased, squashed. I went with exactly the changes you suggested instead of further tweaking things. Let me know if it's all good

File: packages/test-runner-visual-regression/src/visualRegressionPlugin.ts
Comment (web-padawan) [reply packages/test-runner-visual-regression/src/visualRegressionPlugin.ts]: I think that would be a breaking change, while my suggestion isn't.

**Review by web-padawan — COMMENTED** [review]
I have updated the changeset to use `patch` as this does not affect public API of a plugin.

**Review by web-padawan — APPROVED** [review]
LGTM 💯  

Waiting for CI to pass and then I will merge and release a new version.
Thank you for your valuable contributions to this project, I really appreciate it.

**Comment by web-padawan** [comment]
CI failure is unrelated, see #1577. Restarted, hopefully this time it passes 🤞

**Comment by web-padawan** [comment]
Another [unrelated failure](https://github.com/modernweb-dev/web/pull/1657/checks?check_run_id=3494942555), sigh. This also happens from time to time:

```
Failed to execute command test:ci for 3 packages. But we don't know which ones, because concurrently doesn't say.
```

```
[dev-server-import-maps] Error: listen EADDRINUSE: address already in use :::8000
[dev-server-import-maps]     at Server.setupListenHandle [as _listen2] (net.js:1316:16)
[dev-server-import-maps]     at listenInCluster (net.js:1364:12)
[dev-server-import-maps]     at Server.listen (net.js:1450:7)
[dev-server-import-maps]     at internal/util.js:297:30
[dev-server-import-maps]     at new Promise (<anonymous>)
[dev-server-import-maps]     at Server.<anonymous> (internal/util.js:296:12)
[dev-server-import-maps]     at DevServer.start (/home/runner/work/web/web/packages/dev-server-core/dist/server/DevServer.js:36:69)
[dev-server-import-maps]     at TestRunnerServer.start (/home/runner/work/web/web/packages/test-runner-core/dist/server/TestRunnerServer.js:49:30)
[dev-server-import-maps]     at TestRunner.start (/home/runner/work/web/web/packages/test-runner-core/dist/runner/TestRunner.js:56:31)
[dev-server-import-maps]     at Object.startTestRunner (/home/runner/work/web/web/packages/test-runner/dist/startTestRunner.js:49:22) {
[dev-server-import-maps]   code: 'EADDRINUSE',
[dev-server-import-maps]   errno: 'EADDRINUSE',
[dev-server-import-maps]   syscall: 'listen',
[dev-server-import-maps]   address: '::',
[dev-server-import-maps]   port: 8000
[dev-server-import-maps] }
```

Restarted again 🙏

**Review by web-padawan — APPROVED** [review]

**Comment by WickyNilliams** [comment]
> I have updated the changeset to use patch as this does not affect public API of a plugin.

It felt like a new feature (of sorts) to me, that's why i went for a minor version. But that's your call

> Thank you for your valuable contributions to this project, I really appreciate it.

No worries. Thanks for reviewing and merging so quickly! I will probably continue to contribute fixes and improvements as needs arise :) Thanks again

**Comment by web-padawan** [comment]
> It felt like a new feature (of sorts) to me, that's why i went for a minor version. But that's your call

According to semver, during 0.x we use minor versions to indicate breaking changes so I decided to change it.

Sure, keep these fixes coming if you find something that needs improvement 🙇‍♂️

**Comment by WickyNilliams** [comment]
Didn't realise that subtlety of v0.x! Good to know.

**Pull request merged**
